lib/Yield/CrossRelations.cpp

```diff
@@ -0,0 +1,168 @@
+    const auto Location = rank::Function.makeLocation(Function.Entry);
+    Inserter.insert(yield::RelationDescription(Location.toString(), {}));
+  }
+
+  for (const auto &[EntryAddress, ControlFlowGraph] : Metadata) {
+    auto CallLocation = rank::Instruction.makeLocation(EntryAddress,
+                                                       MetaAddress::invalid(),
+                                                       MetaAddress::invalid());
+
+    for (const auto &BasicBlock : ControlFlowGraph) {
+      for (const auto &Edge : BasicBlock.Successors) {
+        if (efa::FunctionEdgeType::isCall(Edge->Type)) {
+          if (const auto &Callee = Edge->Destination; Callee.isValid()) {
+            // TODO: embed information about the call instruction into
+            //       `CallLocation` after efa starts providing it.
+
+            auto Location = rank::Function.makeLocation(Callee).toString();
+            if (auto It = Relations.find(Location); It != Relations.end()) {
+              yield::RelationTarget T(yield::RelationType::IsCalledFrom,
+                                      CallLocation.toString());
+              It->Related.insert(std::move(T));
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+template<typename AddNodeCallable, typename AddEdgeCallable>
+static void convertionHelper(const yield::CrossRelations &Input,
+                             const AddNodeCallable &AddNode,
+                             const AddEdgeCallable &AddEdge) {
+  for (const auto &[LocationString, Related] : Input.Relations)
+    AddNode(LocationString);
+
+  for (const auto &[LocationString, Related] : Input.Relations) {
+    for (const auto &[RelationKind, TargetString] : Related) {
+      switch (RelationKind) {
+      case yield::RelationType::IsCalledFrom:
+        AddEdge(LocationString, TargetString, RelationKind);
+        break;
+
+      case yield::RelationType::Invalid:
+      case yield::RelationType::Count:
+      default:
+        revng_abort("Unknown enum value");
+      }
+    }
+  }
+}
+
+GenericGraph<yield::CrossRelations::Node, 16, true>
+yield::CrossRelations::toGenericGraph() const {
+  GenericGraph<yield::CrossRelations::Node, 16, true> Result;
+
+  using NodeView = decltype(Result)::Node *;
+  std::unordered_map<std::string_view, NodeView> LookupHelper;
+  auto AddNode = [&Result, &LookupHelper](std::string_view Location) {
+    auto *Node = Result.addNode(Location);
+    auto [Iterator, Success] = LookupHelper.try_emplace(Location, Node);
+    revng_assert(Success);
+  };
+  auto AddEdge = [&LookupHelper](std::string_view From,
+                                 std::string_view To,
+                                 yield::RelationType::Values Kind) {
+    using EL = yield::CrossRelations::EdgeLabel;
+    LookupHelper.at(From)->addSuccessor(LookupHelper.at(To), EL{ Kind });
+  };
+  convertionHelper(*this, AddNode, AddEdge);
+
+  return Result;
+}
+
+namespace concepts = pipeline::concepts;
+
+// WIP
```

Comment:
?

Reply:
Sorry, read this like `note to self: following is bunch of templates I'm still unsure about`. I used `WIP` so that `check-conventions` complains about it so that I don't forget to clean the following up.

Reply:
I cleaned up this stuff somewhat and moved it to the `Location.h` header. If you have some suggestions about this, please open a new thread there.